crates/ruff_linter/src/rules/pyupgrade/rules/quoted_annotation.rs

```diff
@@ -87,14 +87,14 @@ pub(crate) fn quoted_annotation(checker: &mut Checker, annotation: &str, range:
     let placeholder_range = TextRange::up_to(annotation.text_len());
     let spans_multiple_lines = annotation.contains_line_break(placeholder_range);
 
-    let tokenizer = SimpleTokenizer::new(annotation, placeholder_range);
-    let last_token_is_comment = matches!(
-        tokenizer.last(),
-        Some(SimpleToken {
-            kind: SimpleTokenKind::Comment,
-            ..
-        })
-    );
+    let last_token_is_comment = checker
+        // Inside an annotation, the `checker.tokens()`
+        // returns the tokens for the parsed annotation.
+        .tokens()
+        // The actual last token will always be a logical newline,
+        // so we check the second to last
+        .get((checker.tokens().len()) - 2)
```

Comment:
The parentheses doesn't seem necessary here.

What would happen if the annotation only contains the comment character i.e., `foo: "#"` which would give `Comment, Newline`? And, similarly for empty annotation like `foo: ""` which would just give `Newline`. I think we should use `saturating_sub` here.

```suggestion
        .get(checker.tokens().len() - 2)
```

Reply:
I don't think either of those cases can occur inside this lint check because it is applied to _parsed_ annotations, and those two examples are syntax errors.

https://github.com/astral-sh/ruff/blob/3125332ec1bc86494feef57c6a6c64d3591a4831/crates/ruff_linter/src/checkers/ast/mod.rs#L2341-L2356

But I agree about the saturating subtraction!